File: src/211. Design Add and Search Words Data Structure.js
export class WordDictionary {
  /**
   * Initialize your data structure here.
   */
  constructor() {
    this.root = {}
  }
  /**
   * @param {string} word
   * @return {void}
   */
  addWord(word) {
    let current = this.root
    for (let char of word) {
      if (!current[char]) current[char] = {}
      current = current[char]
    }
    current.isEnd = true
  }
  /**
   * @param {string} word
   * @return {boolean}
   */
  search(word) {
    return this._searchHelper(word, 0, this.root)
  }
  _searchHelper(word, idx, root) {
    if (typeof root !== 'object') return false
    if (idx === word.length) return !!root.isEnd
    const char = word[idx]
    let found = false
    if (char === '.') {
      for (let nextChar in root) {
        let tmpFound = this._searchHelper(word, idx + 1, root[nextChar])
        found = found || tmpFound
      }
    } else {
      found = this._searchHelper(word, idx + 1, root[char])
    }
    return found
  }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * var obj = new WordDictionary()
 * obj.addWord(word)
 * var param_2 = obj.search(word)
 */

File: src/211. Design Add and Search Words Data Structure.cpp
class TrieNode {
public:
    bool isLeaf = false;
    TrieNode* children[26];
};

class WordDictionary {
private:
    TrieNode* root;
    bool searchHelper(TrieNode* node, string& word, int start) {
        for (int i = start; i < word.size(); ++i) {
            auto c = word[i];
            if (c == '.') {
                for (int j = 0; j < 26; ++j) {
                    if (node->children[j] && searchHelper(node->children[j], word, i + 1)) {
                        return true;
                    }
                }
                return false;
            }
            if (!node->children[c - 'a']) {
                return false;
            }
            node = node->children[c - 'a'];
        }
        return node->isLeaf;
    }
public:
    WordDictionary() {
        root = new TrieNode();
    }
    
    void addWord(string word) {
        auto node = root;
        for (auto c : word) {
            int i = c - 'a';
            if (!node->children[i]) {
                node->children[i] = new TrieNode();
            }
            node = node->children[i];
        }
        node->isLeaf = true;
    }
    
    bool search(string word) {
        return searchHelper(root, word, 0);
    }
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary* obj = new WordDictionary();
 * obj->addWord(word);
 * bool param_2 = obj->search(word);
 */